src/iter/fold_chunks_with.rs

```diff
@@ -0,0 +1,286 @@
+    }
+}
+
+impl<I, U, F> ParallelIterator for FoldChunksWith<I, U, F>
+where
+    I: IndexedParallelIterator,
+    U: Send + Clone,
+    F: Fn(U, I::Item) -> U + Send + Sync,
+{
+    type Item = U;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: Consumer<U>,
+    {
+        bridge(self, consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.len())
+    }
+}
+
+impl<I, U, F> IndexedParallelIterator for FoldChunksWith<I, U, F>
+where
+    I: IndexedParallelIterator,
+    U: Send + Clone,
+    F: Fn(U, I::Item) -> U + Send + Sync,
+{
+    fn len(&self) -> usize {
+        div_round_up(self.base.len(), self.chunk_size)
+    }
+
+    fn drive<C>(self, consumer: C) -> C::Result
+    where
+        C: Consumer<Self::Item>,
+    {
+        bridge(self, consumer)
+    }
+
+    fn with_producer<CB>(self, callback: CB) -> CB::Output
+    where
+        CB: ProducerCallback<Self::Item>,
+    {
+        let len = self.base.len();
+        return self.base.with_producer(Callback {
+            chunk_size: self.chunk_size,
+            len,
+            item: self.item,
+            fold_op: self.fold_op,
+            callback,
+        });
+
+        struct Callback<CB, T, F> {
+            chunk_size: usize,
+            len: usize,
+            item: T,
+            fold_op: F,
+            callback: CB,
+        }
+
+        impl<T, U, F, CB> ProducerCallback<T> for Callback<CB, U, F>
+        where
+            CB: ProducerCallback<U>,
+            U: Send + Clone,
+            F: Fn(U, T) -> U + Send + Sync,
+        {
+            type Output = CB::Output;
+
+            fn callback<P>(self, base: P) -> CB::Output
+            where
+                P: Producer<Item = T>,
+            {
+                self.callback.callback(FoldChunksWithProducer {
+                    chunk_size: self.chunk_size,
+                    len: self.len,
+                    item: self.item,
+                    fold_op: &self.fold_op,
+                    base,
+                })
+            }
+        }
+    }
+}
+
+struct FoldChunksWithProducer<'f, P, U, F>
+where
+    P: Producer,
+{
+    chunk_size: usize,
+    len: usize,
+    item: U,
+    fold_op: &'f F,
+    base: P,
+}
+
+impl<'f, P, U, F> Producer for FoldChunksWithProducer<'f, P, U, F>
+where
+    P: Producer,
+    U: Send + Clone,
+    F: Fn(U, P::Item) -> U + Send + Sync,
+{
+    type Item = F::Output;
+    type IntoIter = FoldChunksWithSeq<'f, P, U, F>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        FoldChunksWithSeq {
+            chunk_size: self.chunk_size,
+            len: self.len,
+            item: self.item,
+            fold_op: self.fold_op,
+            inner: if self.len > 0 { Some(self.base) } else { None },
+        }
+    }
+
+    fn min_len(&self) -> usize {
+        div_round_up(self.base.min_len(), self.chunk_size)
+    }
+
+    fn max_len(&self) -> usize {
+        self.base.max_len() / self.chunk_size
+    }
```

Comment:
Hi, sorry if it's not an actual issue but if I understand correctly `FoldChunks{,With}Producer::max_len` must also use `div_round_up`, otherwise `min_len` may end up larger than `max_len`. For example with the following configuration:

- `self.base.max_len() == 3`
- `self.base.min_len() == 3`
- `self.chunk_size == 2`

then `min_len == 2` and `max_len == 1`.

Reply:
I think you're correct, good catch! I took this from the `chunks::ChunkProducer` impl without properly checking :flushed:

@cuviper: shouldn't the impl in `chunks::ChunkProducer` be doing the same thing?

Reply:
Well, you can already set inconsistent values using `with_min_len` and `with_max_len`, and the way it's currently [documented](https://docs.rs/rayon/latest/rayon/iter/trait.IndexedParallelIterator.html#method.with_max_len) and implemented is that the minimum "wins". The maximum is kind of a soft limit, and we might not be able to satisfy both even when the maximum _is_ more than the minimum.

With chunks, either here or the `chunks` adaptor, it seems messier -- but I think it's correct the way it is. We want to find `n` such that `chunk_size * n` respects the inner limit. So for the minimum, we round `n` up to make sure `chunk_size * n` is never less than the inner minimum, and likewise the maximum should round down so `chunk_size * n` is never more than the inner maximum.

With the example of inner `min == max == 3`, and `chunk_size == 2`, there's no possible way to satisfy this entirely. If we pick `n == 1`, then `n * chunk_size == 2` violates the inner minimum, but with `n == 2` then `n * chunk_size == 4` violates the inner maximum. So, I think it's simplest if `min_len` and `max_len` go ahead and compute their ideal limits independently, and then the rest of the code that *uses* these limits will favor the minimum, as I mentioned before.